FIxes #8951

acolytec3 opened the pull request:
fixes #8951 

### Summary

Fixes both scenarios from #8951 where balance is not updated after reconnecting to network

### Testing notes
Tested on Oneplus 6T

#### Areas that maybe impacted
Portfolio Balance view
Account Balance view

##### Functional
- wallets
- account
### Steps to test

Scenario 1
- Open Status
- Turn on airplane mode
- Restore any account with seed phrase or keycard and login
- Verify account balance shows as "..."
- Turn off airplane mode and wait up to 30 seconds
- Balance should update

Scenario 2
- Open Status
- Login with any account
- Turn on airplane mode
- Send transaction from another account and wait for confirmation on Etherscan
- Turn off airplane mode and wait ~30 seconds to a minute
- Balance should update


status: ready

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b06140b4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-9157/1/) | 2019-10-10 11:23:00 | ~11 min | `android-e2e` | [:package:`apk`](https://status-im.ams3.digitaloceanspaces.com/StatusIm-191010-111128-b06140-e2e-x86.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A0&chl=https%3A%2F%2Fstatus-im.ams3.digitaloceanspaces.com%2FStatusIm-191010-111128-b06140-e2e-x86.apk)|

acolytec3 committed:
Fix network status change logic and #8951

acolytec3 commented:
@yenda I fixed a secondary logic bug in the `handle-network-info-change` effect so it only calls `change-network-status` when the network status actually changes instead of every single time there's a new `network-inifo-changed` event.  As such, the current state of the PR is that `network/status` is now correctly being updated when there's an actual network change.  I added a call to `update-balances` every time  the network status changes to `:online` and this does solve the base problem of the account balances not being nil.  

That said, this has fallen back on the "bad" solution you identified in the issue discussion (i.e. update balances every time the app detects that we've reconnected to the network).  The reason being, within the `change-network-status` effect, I can't come up with a simple way to grab the balance field for any of the individual accounts since the individual account-portfolio-balance subscriptions are set at run-time like `:[account-portfolio-balance "0x1a2b3c..."]` and we aren't passing accounts to `change-network-status`.  Am I over thinking this?  It seems like there should be an elegant solution to just do a `select-keys [:account-portfolio-value]` or something similar and then check if the value is '..." before calling update-balances but I think my newness to Clojurescript is getting in the way of me finding it.

yenda commented on src/status_im/network/net_info.cljs:
```diff
@@ -7,12 +7,14 @@
   [{:keys [db] :as cofx} is-connected?]
   (fx/merge cofx
             {:db (assoc db :network-status (if is-connected? :online :offline))}
+            (if is-connected? (wallet/update-balances nil))
```
You can use when here instead of if.

Also subscriptions are for views, in events you get the values in db so you can check the balances from 'db' here, check how the balance is checked in the subscription to get it from db. I'm on mobile but if you need more details I can check and point you to the right place

acolytec3 commented:
Fixed the if to when.

I think I've identified a possibly better route to identifying when to call the `update-balances` effect which centers on checking for the`balance-update` error. My first attempt didn't quite work but I think this is the right approach. More to come tomorrow.

acolytec3 commented:
@yenda This PR should now be ready for review.  The additional change I made is in `change-network-status` where it first checks to see if the change is from off-line to online and then looks to see if any accounts have been loaded to the wallet in the db.  If not, it then calls update-balances.  This solves scenario 1 since the accounts are flushed from the wallet in the DB every time you sign out and the accounts aren't loaded if you login to the app while `network-status` is set to offline.  Scenario 2 is covered by status-go updates.  If you go offline while staying logged into the app, when it comes back online, the only time your balance will have changed is if you get a new transfer and this also calls update-balances.

I'm not sure why all the builds are failing on Jenkins right now but it works locally.

acolytec3 commented:
@yenda Circling back on this.  Are you good with this approach for determining when to call `update-balances`?

yenda reviewed (approved):
@acolytec3 is it fixed with current changes? Code wise it looks good

acolytec3 commented:
@yenda From my testing, yes. Would like for QA to confirm that behaviour is
as expected.

On Thu, Oct 17, 2019, 5:39 PM yenda <notifications@github.com> wrote:

> *@yenda* approved this pull request.
>
> @acolytec3 <https://github.com/acolytec3> is it fixed with current
> changes? Code wise it looks good
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/9157?email_source=notifications&email_token=AEENFXG6L7HUP3RSD2QPBODQPDLPTA5CNFSM4I7LMV62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIL75TI#pullrequestreview-303562445>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEENFXBFNTFPFRUSPJ5AP63QPDLPTANCNFSM4I7LMV6Q>
> .
>

acolytec3 commented:
@churik Any chance you can test this and validate that my changes resolve the issue?  Thanks!

churik commented:
@acolytec3 please, rebase your PR to current develop. Thank you.

acolytec3 commented:
@churik Done.

churik commented:
@acolytec3 
**Issue 1. Error right after launch app**
Logcat:
```
10-24 10:11:57.745 28097 28137 I ReactNativeJS: NAME: "Error"
10-24 10:11:57.745 28097 28137 I ReactNativeJS: MESSAGE: Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when trying to update settings
10-24 10:11:57.745 28097 28137 I ReactNativeJS: 
10-24 10:11:57.745 28097 28137 I ReactNativeJS: Error: Please shake the phone to report this error and restart the app. multiaccount is currently empty, which means something went wrong when trying to update settings
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3327265)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3326550)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273113)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at pu (address at index.android.bundle:1:2152734)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3053868)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3037147)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3067374)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273007)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3373969)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3372906)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273113)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at pu (address at index.android.bundle:1:2152734)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3053868)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3037147)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3067374)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273007)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3598197)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3597712)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273113)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at pu (address at index.android.bundle:1:2152734)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3053868)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3037147)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3067374)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3273007)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3599584)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:3599650)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:2273402)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at WQn (address at index.android.bundle:1:2220699)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at G1n (address at index.android.bundle:1:2268566)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at X1n (address at index.android.bundle:1:2271444)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at Q1n (address at index.android.bundle:1:2268811)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:2271141)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at Z1n (address at index.android.bundle:1:2269059)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at n (address at index.android.bundle:1:2268947)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at apply (native)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:4297795)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at y (address at index.android.bundle:1:4296886)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at C (address at index.android.bundle:1:4297088)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at callImmediates (address at index.android.bundle:1:4298888)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at value (native)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at anonymous (address at index.android.bundle:1:4187305)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at value (address at index.android.bundle:1:4188042)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at value (address at index.android.bundle:1:4187216)
10-24 10:11:57.745 28097 28137 I ReactNativeJS:     at value (native)
10-24 10:11:57.745 28097 28137 I ReactNativeJS: 
10-24 10:11:57.745 28097 28137 I ReactNativeJS: ***
```
The same error after going to offline.

**Issue 2: updating balance is working only for overall balance, not for individual account**

Consider the following scenario:
- restore account with positive balance on several accounts (let's say on 2) in multiaccount 
- go to offline
- create 2 account
- go to online

**Expected result:** balance for second account is fetched
**Actual result:** balance is not fetched

Another scenario:
- restore account with positive balance on several accounts
- send some ETH (I was on Ropsten test network) from one account to another
- go to offline 
- in ~5 min go to online 
**Expected result:** balance of Status account and second account are updated
**Actual result:** balances are the same; updated after relogin.

acolytec3 commented:
@churik I've recreated scenario 1 that outline above and pushed a fix for it.  I'm not able to recreate scenario 2.  When I login with a multiaccount where 2 accounts have a positive balance, send a transaction from one to the other, immediately go to airplane mode before the transaction is mined, wait several minutes, and then turn airplane mode off and restore network connectivity, I see the confirmed transaction appear in the second account and the individual balances update accordingly.  I'm also no longer seeing the initial error that you noted.

Can you retest now?

acolytec3 commented:
@yenda I updated the logic for deciding when to update-balances after coming online and it now compares the number of accounts in the wallet versus the number of accounts under the multiaccount and calls `update-balances` if not equal.  The logic here is that if the count is different, that indicates that new accounts were created in multiaccount while disconnected.  Should resolve the edge cases that @churik noted in testing.

Serhy commented:
Tested latest builds (https://i.diawi.com/xhVHPa and https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-191024-104159-22fc4e-pr-universal.apk) with iOS and Android 8.1
There are no issues with all scenarios https://github.com/status-im/status-react/pull/9157#issuecomment-545810601 or from issue description. 
Balances (ETH and token) are updated on any wallet view: main Wallet screen or particular account view

Thank you, @acolytec3

acolytec3 commented:
@serhy Awesome!  Do you still need me to squash and update the PR?  Sorry, forgot to do that when I pushed the new commit.

Serhy commented:
> @Serhy Awesome! Do you still need me to squash and update the PR? Sorry, forgot to do that when I pushed the new commit.

Yes, could you squash commits, please?

acolytec3 commented:
Yep, will do it later today.

acolytec3 commented:
@Serhy squash complete.
